tensorflow/compiler/mlir/hlo/lib/Transforms/buffer_packing.cc

```diff
@@ -0,0 +1,476 @@
+          gaps.push_back(
+              std::make_pair(UseInterval(oldStart, allocToPack.firstUse - 1),
+                             freeContiguousMemory));
+      }
+      return true;
+    }
+    return false;
+  }
+
+  /// Aggreagtes the allocation informations of the allocs.
+  size_t computeAllocationInfos(AllocInfoList &allocInfos,
+                                const UserangeAnalysis &userangeAnalysis,
+                                const mlir::BufferPlacementAllocs &allocs) {
+    // Create allocInformations and store them in allocInfos.
+    size_t maxUserangeId = 0;
+
+    for (auto &allocEntry : allocs) {
+      Value v = std::get<0>(allocEntry);
+      auto userangeIntervals = userangeAnalysis.getUserangeInterval(v);
+
+      if (!userangeIntervals)
+        continue;
+
+      // Computes the userange id of the allocation.
+      size_t allocUserangeId = userangeAnalysis.computeId(v, v.getDefiningOp());
+
+      // Computes the last use of the allocated buffer.
+      size_t lastUse = std::prev((*userangeIntervals.getValue()).end())->end;
+
+      // Computes the first use of the allocated buffer.
+      size_t firstUse = (*userangeIntervals.getValue()).begin()->start;
+
+      // Computes the number of aligend segments of the buffer.
+      size_t numSegments = computeAlignedSegments(v);
+      maxUserangeId = std::max(maxUserangeId, lastUse);
+      allocInfos.emplace_back(v, allocUserangeId, firstUse, lastUse,
+                              numSegments, 0, userangeIntervals.getValue());
+    }
+
+    // If the window size is zero we are ready.
+    if (windowSize == 0)
+      return maxUserangeId;
+    // Sorts the allocation informations to compute the window id.
+    std::sort(allocInfos.begin(), allocInfos.end(),
+              [](const AllocationInfo &a, const AllocationInfo &b) {
+                return a.allocUserangeId < b.allocUserangeId;
+              });
+
+    // resize window id
+    size_t windowId = 0;
+    size_t lastAllocUserangeId = 0;
+    for (auto &allocationInfo : allocInfos) {
+      if (allocationInfo.allocUserangeId > lastAllocUserangeId + windowSize)
+        ++windowId;
+
+      lastAllocUserangeId = allocationInfo.allocUserangeId;
+      allocationInfo.windowId = windowId;
+    }
+    return maxUserangeId;
+  }
+};
+
+/// Template to reuses already allocated buffer to save allocation operations.
+class BufferPacking : BufferPlacementTransformationBase {
+
+public:
+  template <typename StrategyT>
+  BufferPacking(Operation *op, StrategyT strategy)
+      : BufferPlacementTransformationBase(op),
+        userangeAnalysis(op, allocs, aliases), dominators(op) {
+    DenseMap<size_t, size_t> bufferGen;
+    std::vector<PackedInfo> packingInfos;
+    strategy.optimze(allocs, userangeAnalysis, packingInfos, bufferGen);
+
+    DenseMap<size_t, Value> generatedBuffers;
+
+    // Find common dominators.
+    Block *block = findAllocationsDominator(packingInfos);
```

Comment:
Should this be per packed allocation?

Reply:
We pritify the optimize and resolved it.